### PR TITLE
Add tests for token api

### PR DESCRIPTION
**Pull request by graemecode**
* Adds some tests for Token API, including:
  * transferAsync checks that sender has sufficient balance, throw if not the case
  * transferFromAsync checks that sender has sufficient balance and allowance
  * getBalanceAsync throws if contract does not implement ERC20 interface
  * setProxyAllowance throws if contract does not implement the ERC20 interface
  * setUnlimitedProxyAllowance throws if contract does not implement the ERC20 interface
  * getProxyAllowanceAsync throws if contract does not implement the ERC20 interface
* Removes hack to validate JSONSchema of BigNumbers, by adding custom format validator

### CODE REVIEW

File: src/invariants/token.ts [hunk 1]
```diff
@@ -1,7 +1,24 @@
...
+import * as singleLineString from "single-line-string";
+
+export const TokenAssertionErrors = {
+    MISSING_ERC20_METHOD: () => singleLineString`Contract does not implement ERC20 interface.`,
+};
+
 export class TokenAssertions {
+    // Throws if the given candidateContract does not respond to some methods from the ERC20 interface.
+    // TODO: This could be made more complete by comparing the ERC20 interface to the candidate's properties.
+    public async implementsERC20(candidate: any): Promise<void> {
```
Comment (nadavhollander) [inline src/invariants/token.ts]: `candidate` should be of type `ERC20Contract`

File: src/invariants/token.ts [hunk 2]
```diff
@@ -1,7 +1,24 @@
 import { BigNumber } from "bignumber.js";
 import { ERC20Contract } from "../wrappers";
 
+import * as singleLineString from "single-line-string";
+
+export const TokenAssertionErrors = {
+    MISSING_ERC20_METHOD: () => singleLineString`Contract does not implement ERC20 interface.`,
```
Comment (nadavhollander) [inline src/invariants/token.ts]: Let's add an `address` parameter and specify _which_ contract does not implement the required interface. 

i.e. `Contract at address ${contractAddress} does not implement ERC20 interface `

File: src/schemas/schema_validator.ts [hunk 3]
```diff
@@ -21,13 +22,19 @@ export class SchemaValidator {
...
-    // In order to validate a complex JS object using jsonschema, we must replace any complex
-    // sub-types (e.g BigNumber) with a simpler string representation. Since BigNumber and other
-    // complex types implement the `toString` method, we can stringify the object and
-    // then parse it. The resultant object can then be checked using jsonschema.
     public validate(instance: any, schema: Schema): ValidatorResult {
-        const jsonSchemaCompatibleObject = JSON.parse(JSON.stringify(instance));
-        return this._validator.validate(jsonSchemaCompatibleObject, schema);
+        Validator.prototype.customFormats.BigNumber = function(input) {
+            const regex = RegExp("^\\d+(\\.\\d+)?$");
+            // This allows undefined inputs, e.g. salt is not present sometimes.
```
Comment (nadavhollander) [inline src/schemas/schema_validator.ts]: This feels like something a bit code-smellish that shouldn't be necessary -- why are format validations getting run on optional parameters?

File: src/invariants/token.ts [hunk 4]
```diff
@@ -1,7 +1,47 @@
...
+import * as singleLineString from "single-line-string";
+
+export const TokenAssertionErrors = {
+    MISSING_ERC20_METHOD: () =>
+        singleLineString`Contract does not implement ERC20 interface.`,
+};
+
 export class TokenAssertions {
+    // Throws an error if the given candidateContract does not implement the ERC20 interface.
+    public async implementsERC20(candidate: any): Promise<void> {
```
Comment (nadavhollander) [inline src/invariants/token.ts]: `candidate` should be of type `ERC20Contract` for type safety

File: src/invariants/token.ts [hunk 5]
```diff
@@ -1,7 +1,47 @@
...
+                'approve',
+                'allowance'
+            ];
+
+            specMethods.map((method: string) => {
+                if (!candidate.hasOwnProperty(method)) {
+                    throw new Error(TokenAssertionErrors.MISSING_ERC20_METHOD(method));
+                }
+            });
+        */
```
Comment (nadavhollander) [inline src/invariants/token.ts]: Interesting -- I think doing the basic checks you've put in is sufficient -- let's add some sort of `NOTE` or `HACK` comment to explicate that we're only testing for a subset of the ERC20 methods.

File: src/schemas/schema_validator.ts (hunk 3 above)
Comment (nadavhollander) [reply src/schemas/schema_validator.ts]: If this is something unavoidable I'm okay with it, but my impression was that json schema validations _in general_ don't run on optional parameters unless they're present.

File: src/invariants/token.ts [hunk 6]
```diff
@@ -1,7 +1,47 @@
 import { BigNumber } from "bignumber.js";
 import { ERC20Contract } from "../wrappers";
 
+import * as singleLineString from "single-line-string";
+
+export const TokenAssertionErrors = {
+    MISSING_ERC20_METHOD: () =>
+        singleLineString`Contract does not implement ERC20 interface.`,
```
Comment (nadavhollander) [inline src/invariants/token.ts]: Let's add a parameter for `contractAddress`, i.e.

`Contract at address ${contractAddress} does not implement ERC20 interface.`

**Comment by graemecode** [comment]
♻️

File: __test__/integration/token_api.spec.ts [hunk 7]
```diff
@@ -160,6 +184,44 @@ describe("Token API (Integration Tests)", () => {
...
+                            SPENDER,
+                            RECIPIENT,
+                            new BigNumber(10),
+                            { from: OPERATOR },
+                        ),
+                    ).rejects.toThrow(TokenAPIErrors.INSUFFICIENT_SENDER_BALANCE());
+                });
+            });
+
+            describe("sender has insufficient allowance", async () => {
```
Comment (nadavhollander) [inline __test__/integration/token_api.spec.ts]: Please change description to `"spender has given operator insufficient allowance"`

File: src/apis/token_api.ts [hunk 8]
```diff
@@ -1,18 +1,32 @@
...
 import { Web3Utils } from "../../utils/web3_utils";
 import { ContractsAPI } from "./";
 import { BigNumber } from "bignumber.js";
+import { Assertions } from "../invariants";
 import { TxData } from "../types";
 
 const TRANSFER_GAS_MAXIMUM = 70000;
 
+export const TokenAPIErrors = {
+    INSUFFICIENT_SENDER_BALANCE: () =>
```
Comment (nadavhollander) [inline src/apis/token_api.ts]: Let's add an argument to this that specifies the spender's address -- i.e. `"Sender with address ${spenderAddress} does not have sufficient balance..."`

**Comment by graemecode** [comment]
♻️

**Pull request merged**